### ISSUE TITLE
Overlay video window on macOS Sierra (PiP)

### ISSUE THREAD

**Issue by herrbischoff** [mpv-player/mpv#3612]
In Sierra there is a new feature of an overlay video window that's resizable and stays put on all spaces. If it is at all possible to implement this behaviour in mpv, I would very much appreciate it being added. I am aware of an already available overlay function. However, this overlaid window stays on the space it was put, not floating above different spaces.

**Comment by joshdekock** [comment]
Could you point to this function? As far as I'm aware it looks like a safari-only feature.

**Comment by herrbischoff** [comment]
It indeed appears to be limited to Safari and iTunes for the time being. I was hoping someone might have traversed the developer docs deeper than I have and found something to make it work with other applications.

**Comment by joshdekock** [comment]
Yes, I think it's a private API for now. Definitely something which may be possible in the future but at this point in time it doesn't seem so (without reimplementing it).

**Comment by sashimiman** [comment]
Hi, I accidentally opened a new feature request about PiP. Here's info which may be relevant :

https://developer.apple.com/library/content/documentation/WindowsViews/Conceptual/AdoptingMultitaskingOniPad/QuickStartForPictureInPicture.html

kind regards

**Comment by Akemi** [comment]
some additional info. we would also need [NSWindowCollectionBehaviorFullScreenAuxiliary](https://developer.apple.com/reference/appkit/nswindowcollectionbehavior/nswindowcollectionbehaviorfullscreenauxiliary)  in combination with `NSWindowCollectionBehaviorCanJoinAllSpaces`.

**Comment by SteveJobzniak** [comment]
@Akemi That's incredible. So it seems like there is no need to hack into the private PiP framework. Just setting "[window can appear on all spaces](https://developer.apple.com/reference/appkit/nswindowcollectionbehavior/nswindowcollectionbehaviorcanjoinallspaces?language=objc)" and "[window can appear on top of fullscreen apps](https://developer.apple.com/reference/appkit/nswindowcollectionbehavior/nswindowcollectionbehaviorfullscreenauxiliary)" sounds like a brilliant and pretty easy solution, if that's really all it takes.

And as you said (https://github.com/mpv-player/mpv/commit/a8347eb9ba9e09177da50592fde7f3ae7261ce59#commitcomment-20239490), it _may_ also need some window snapping logic if you want to go the _whole_ way towards emulating PiP, but that's _nowhere_ near as important as just being able to _see_ the mpv window on all spaces. That's the single remaining benefit of Apple's PiP over mpv when working.

If this is added, I would prefer it to be a triggerable Lua/Keybinding command, perhaps by setting one or two new mpv properties: "cocoa-canjoinallspaces" and "cocoa-showonfullscreenspace" (with some nicer names). Then people could trigger it by setting those properties either in mpv.conf or via Lua/keybinds. And they could opt in to "always show on all spaces, except fullscreen spaces", etc. Which would be very impressive.

**Comment by SteveJobzniak** [comment]
I found some additional info:

https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/WinPanel/Articles/SettingWindowCollectionBehavior.html

> **Spaces Collection Behavior**
> 
> There are three options that can be set for a window’s Spaces collection behavior. The default is NSWindowCollectionBehaviorDefault, which allows the window to be associated with one space at a time. The second option is NSWindowCollectionBehaviorCanJoinAllSpaces. This option causes the window to appear on all spaces, like the menu bar. The third option is NSWindowCollectionBehaviorMoveToActiveSpace. This causes the window to switch to the active space when it is made active. Only one of these options may be used at a time.
> 
> If a window is currently associated with the active space, isOnActiveSpace returns YES. Otherwise, it returns NO. Additionally, you can get an array of the window numbers of windows on one or all spaces using the method windowNumbersWithOptions: and specified your desired options. The possible options are specified by NSWindowNumberListOptions.
> 
> **Exposé Collection Behavior**
> 
> There are also three options that can be set for a window’s Exposé collection behavior. If a window has a window level of NSNormalWindowLevel, the default behavior is NSWindowCollectionBehaviorManaged, which causes the window to participate in both Spaces and Exposé. NSWindowCollectionBehaviorTransient causes the window to float in Spaces and be hidden in Exposé. This is the default behavior if the window level is not NSNormalWindowLevel. The final option is NSWindowCollectionBehaviorStationary, which causes the window to be unaffected by Exposé; i.e. it stays visible and does not move, like the desktop window. Only one of these options may be used at a time.

**Comment by nguyenhimself** [comment]
Hate to be that guy, but has there been any progress on this feature?

I was reminded of this because[ IINA](https://lhc70000.github.io/iina), a macOS front-end for mpv, has added Picture-in-Picture, and well, it's kinda nice :) I still prefer the mainline mpv though.

**Comment by Akemi** [comment]
as far as it concerns me i didn't work on it (yet). IINA also uses a private API which isn't an option for mpv.

**Comment by SteveJobzniak** [comment]
@nguyenhimself Wow, that IINA project is everything that BakaMPlayer (the other, older mpv frontend for Mac) is not.

I would switch to IINA if it had 100% compatibility with ~/.config/mpv and all user scripts and custom keybindings etc. Do you know if that's the case? Or do they only use their own locked-down settings system?

Something like IINA is great for my girlfriend who doesn't remember the keyboard shortcuts I use in mpv.

**Comment by Akemi** [comment]
mpv itself isn't interested in adding a proper GUI. it encouraged others to use libmpv and do such things. IINA uses all the things mpv also supports, though not everything works properly so far. one thing i know of is display-resamle+interpolation.

**Comment by nguyenhimself** [comment]
@SteveJobzniak You can supposedly point IINA to your own config file, but in reality, my rather custom config file made IINA hang and crash. Because IINA's log file is useless for crash debugging, I can't be bothered enough to comment out each option in my config file to find out what tripped up IINA.

Frankly I'm perfectly happy with mpv's UI. Picture-In-Picture is the only macOS-native thing that I wish mpv could copy from IINA, but as @Akemi had explained, I'm willing to wait.

**Comment by Akemi** [comment]
@nguyenhimself if i had to guess it's probably your `vo=opengl-hq` line.

**Comment by nguyenhimself** [comment]
@Akemi It's likely that you're right. I'll keep that in mind next time I download and install IINA :)

**Comment by SteveJobzniak** [comment]
@Akemi Thanks for the explanation. That's actually a cool idea. Letting others make more advanced GUIs via libmpv.

@nguyenhimself I think I may prefer the classic mpv GUI too, and it's easy enough for even non-geeks to remember "q" to quit and to just use the mouse to click the OSD seek-bar manually. But the problem is when advanced things like playlists, subtitle selection, audio track, audio delay, etc needs to be done. It really confuses everyone I know (not myself). So I'll be trying out IINA. Thanks both of you for the information! <3 I will comment out opengl-hq and try it soon.

**Comment by herrbischoff** [comment]
It's been three years since the last comment on this issue. It appears unlikely this feature ever gets implemented and with IINA there is now a macOS "native" player for those who seek it. Closing.

**Comment by Akemi** [comment]
since i am very much interested in this feature myself i will reopen this feature request. i am still hoping for a none private framework that 'just works' solution, though this will probably not happen.
